lib/resty/core/socket.lua

```diff
@@ -107,17 +133,132 @@ local function setoption(cosocket, option, value)
                                                        err,
                                                        errlen)
     if rc ~= FFI_OK then
-        return nil, ffi_string(err, errlen[0])
+        return nil, ffi_str(err, errlen[0])
+    end
+
+    return true
+end
+
+
+local errmsg = base.get_errmsg_ptr()
+local session_ptr = ffi_new("void *[1]")
+local server_name_str = ffi_new("ngx_str_t[1]")
+local openssl_error_code = ffi_new("int[1]")
+
+
+local function setclientcert(cosocket, cert, pkey)
+    if not cert and not pkey then
+        cosocket.client_cert = nil
```

Comment:
better use number instead of a string as key index here? since array has better performance than hash.
we can reach it just like `SOCKET_CTX_INDEX`, right?

Reply:
done, thanks.